finagle-core/src/main/scala/com/twitter/finagle/dispatch/ServerDispatcher.scala

```diff
@@ -93,22 +133,37 @@ abstract class GenSerialServerDispatcher[Req, Rep, In, Out](trans: Transport[In,
 
 /**
  * Dispatch requests from transport one at a time, queueing
- * concurrent requests.
+ * concurrent requests. Transport errors are considered fatal; 
+ * the service will be released after any error.
  *
- * Transport errors are considered fatal; the service will be
- * released after any error.
+ * @tparam Req the `Request` type for the [[ServerDispatcher]]
+ * @tparam Rep the `Response` type for the [[ServerDispatcher]]
+ * @param trans a [[com.twitter.finagle.transport.Transport Transport]]
+ * @param service 
+ * @param init a [[ServerDispatcherInitializer]] used to init tracing
  */
 class SerialServerDispatcher[Req, Rep](
     trans: Transport[Rep, Req],
-    service: Service[Req, Rep])
-    extends GenSerialServerDispatcher[Req, Rep, Rep, Req](trans) {
+    service: Service[Req, Rep],
+    init: ServerDispatcherInitializer)
+    extends GenSerialServerDispatcher[Req, Rep, Rep, Req](trans, init) {
 
   trans.onClose ensure {
     service.close()
   }
 
-  protected def dispatch(req: Req, eos: Promise[Unit]) =
-    service(req) ensure eos.setDone()
+  protected def dispatch(req: Req, eos: Promise[Unit]) = init.fReq(req) match {
+    case Some(traceId) => Trace.letTracerAndId(init.tracer, traceId) {
+      Trace.record(Annotation.WireRecv)
+      service(req) ensure eos.setDone()
+    }
+    case None          => service(req) ensure eos.setDone()
+  }
 
-  protected def handle(rep: Rep) = trans.write(rep)
+  protected def handle(rep: Rep) = init.fRep(rep) match {
```

Comment:
Yeah, I wasn't thinking clearly here - we use some http services internally that we would like to have the ID embedded, but those would

- Never have a different ID that than the request
- Wouldn't be affected by the SerialServerDispatcher anyways

Thinking back to a month ago, I originally envisioned the `ServerDispatcherInitializer` as a way to tell the Dispatcher whether you wanted any traces returned by making `fRep: (a: Any) => TraceId` optional. That's probably too cute by half.